INFRA-438 Handle observers not having error handling

rnicoll opened the pull request:
When the RPC client connection is closed, it notifies observers using onError(), which may not be the correct approach (TBD) but changing this is a much more invasive change. Where observers do not subscribe to error notifications, this is reflected to the calling client by an exception thrown.

This change catches that exception and lots it as debug rather an error level.

Commit:
INFRA-438 Handle observers not having error handling

When the RPC client connection is closed, it notifies observers using onError(),
which may not be the correct approach (TBD) but changing this is a much more invasive change.
Where observers do not subscribe to error notifications, this is reflected to the calling
client by an exception thrown.

This change catches that exception and lots it as debug rather an error level.

yift-r3 commented on client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt:
```diff
@@ -455,10 +456,17 @@ internal class RPCClientProxyHandler(
-                    log.error("Unexpected exception when RPC connection failure handling", e)
+                val observable: UnicastSubject<Notification<*>>? = observablesMap[key]
+                // Notify listeners of the observables that the connection is being terminated.
+                observable?.also {
```
Do we actually need the also?

rnicoll replied:
I'd used it as a way of handling nullability, the wrapped code won't be run if the observable doesn't exist.

That said your suggestion of iterating both key & value together is a lot neater.

yift-r3 commented on client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt:
```diff
@@ -455,10 +456,17 @@ internal class RPCClientProxyHandler(
         val observablesMap = observableContext.observableMap.asMap()
         observablesMap.keys.forEach { key ->
```
Maybe 
```
observablesMap.forEach { (key, observable) ->
```
And then we don't need the getter. Also, we can filter for null observables.

dimosr commented on client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt:
```diff
@@ -455,10 +456,17 @@ internal class RPCClientProxyHandler(
+                        it.onError(ConnectionFailureException())
+                    } catch(ex: OnErrorNotImplementedException) {
+                        // Indicates the observer does not have any error handling.
+                        log.debug("Closed connection on observable $observable whose observers have no error handling.")
```
I am not sure about the DEBUG-level here, should we at least keep it as a WARN-level log to inform users in some way that if they want to react to the connection closing event they should register an `onError` handler?

rnicoll replied:
Changed to `warn`

sotr3 commented:
Hi Ross,
I tested this patch. I am getting another exception now but only for  startNodesInProcess = true. 
Out of process completes without errors or exceptions: 
 io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:16500
Caused by: java.net.ConnectException: Connection refused
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method) ~[?:1.8.0_231]
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717) ~[?:1.8.0_231]
	at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:330) ~[netty-transport-4.1.46.Final.jar:4.1.46.Final]
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:334) ~[netty-transport-4.1.46.Final.jar:4.1.46.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:702) ~[netty-transport-4.1.46.Final.jar:4.1.46.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650) ~[netty-transport-4.1.46.Final.jar:4.1.46.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576) ~[netty-transport-4.1.46.Final.jar:4.1.46.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) ~[netty-transport-4.1.46.Final.jar:4.1.46.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) ~[netty-common-4.1.46.Final.jar:4.1.46.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.46.Final.jar:4.1.46.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.46.Final.jar:4.1.46.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_231]

Commit:
Resolve Detekt failures

Commit:
Move the observable closing logic to a separate method

dimosr commented on client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt:
```diff
@@ -142,6 +143,19 @@ internal class RPCClientProxyHandler(
+                observable.onError(ConnectionFailureException())
+            } catch (ex: OnErrorNotImplementedException) {
+                // Indicates the observer does not have any error handling.
+                log.warn("Closed connection on observable $observable whose observers have no error handling.")
```
what is the `$observable` part going to print - will this contain something that the user can use for correlation purposes (e.g. observable ID)?

yift-r3 replied:
The warning looks like:
`[WARN] 09:20:50,884 [Time-limited test] internal.RPCClientProxyHandler. - Closed connection on observable rx.subjects.UnicastSubject@49b6bb49 whose observers have no error handling.`

- This is unhelpful
- It's really hard to get the subscriber from the subscription (we would need to change rx code).
- It looks like subscribing without error handling is a valid use case that we have a lot in the code.

I think we should ignore this exception altogether.

rnicoll replied:
Dropped it down to `trace` level just in case, and removed the `observable` print

Commit:
Reduce error log level

Change log level from warn to trace as there's not a lot that can be done with discovering there's no
error handling, at connection close time. Also removed the observable reference, as it doesn't print anything
useful.

vkolomeyko commented on client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt:
```diff
@@ -142,6 +143,19 @@ internal class RPCClientProxyHandler(
+                observable.onError(ConnectionFailureException())
+            } catch (ex: OnErrorNotImplementedException) {
+                // Indicates the observer does not have any error handling.
+                log.trace("Closed connection on observable whose observers have no error handling.")
```
Minor point, I would put log line as:
```
log.debug { "Closed connection on observable whose observers have no error handling." }
```
I thing `debug` level is good enough for the purpose, also passing lambda is preferred when dealing with log levels that will rarely be enabled during CI runs.

rnicoll replied:
Lambda shouldn't make a difference, should it, given this is a constant string? There's no code executed either way, I mean.

Will change to debug

vkolomeyko replied:
Indeed, in this instance lambda or not will not make a difference.
It is just way too often people concatenating strings which never get logged with `debug`, so I  find it a good habit/style to always use lambda with `debug` for consistency.

mnesbit replied:
Anything at more detail than INFO MUST be using lambdas in our code please! The log string buffers are the No 1 top reason for out of memory in production system.

Commit:
Switch log level to debug

Commit:
Switch to using a lambda

rnicoll commented:
Retesting after https://github.com/corda/corda/pull/6556
